From e6923b32a616e0b16dd67305354de87b87b94e07 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:06:43 -0400
Subject: [powerpc] powernv: Create opal sysfs directory

Message-id: <1410545655-205645-175-git-send-email-dzickus@redhat.com>
Patchwork-id: 94302
O-Subject: [RHEL7 PATCH 174/626] powerpc/powernv: Create opal sysfs directory
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 6f68b5e2c6c04e9cf0e3074f884da36957ce9aae
Author: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
Date:   Tue Aug 27 15:09:52 2013 +0530

    powerpc/powernv: Create opal sysfs directory

    Create /sys/firmware/opal directory. We wil use this
    interface to fetch opal error logs, firmware update, etc.

    Signed-off-by: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/opal.h b/arch/powerpc/include/asm/opal.h
index 4cc33ba..ee0efd2 100644
--- a/arch/powerpc/include/asm/opal.h
+++ b/arch/powerpc/include/asm/opal.h
@@ -601,6 +601,9 @@ typedef struct oppanel_line {
  uint64_t  line_len;
 } oppanel_line_t;
 
+/* /sys/firmware/opal */
+extern struct kobject *opal_kobj;
+
 /* API functions */
 int64_t opal_console_write(int64_t term_number, __be64 *length,
       const uint8_t *buffer);
diff --git a/arch/powerpc/platforms/powernv/opal.c b/arch/powerpc/platforms/powernv/opal.c
index 329924f..3a30a34 100644
--- a/arch/powerpc/platforms/powernv/opal.c
+++ b/arch/powerpc/platforms/powernv/opal.c
@@ -18,11 +18,15 @@
 #include <linux/interrupt.h>
 #include <linux/notifier.h>
 #include <linux/slab.h>
+#include <linux/kobject.h>
 #include <asm/opal.h>
 #include <asm/firmware.h>
 
 #include "powernv.h"
 
+/* /sys/firmware/opal */
+struct kobject *opal_kobj;
+
 struct opal {
  u64 base;
  u64 entry;
@@ -376,6 +380,17 @@ static irqreturn_t opal_interrupt(int irq, void *data)
  return IRQ_HANDLED;
 }
 
+static int opal_sysfs_init(void)
+{
+ opal_kobj = kobject_create_and_add("opal", firmware_kobj);
+ if (!opal_kobj) {
+  pr_warn("kobject_create_and_add opal failed\n");
+  return -ENOMEM;
+ }
+
+ return 0;
+}
+
 static int __init opal_init(void)
 {
  struct device_node *np, *consoles;
@@ -421,6 +436,10 @@ static int __init opal_init(void)
        " (0x%x)\n", rc, irq, hwirq);
   opal_irqs[i] = irq;
  }
+
+ /* Create "opal" kobject under /sys/firmware */
+ rc = opal_sysfs_init();
+
  return 0;
 }
 subsys_initcall(opal_init);
-- 
1.7.1